Mathlib/RingTheory/AdjoinRoot.lean

```diff
@@ -401,56 +408,138 @@ theorem noZeroSMulDivisors_of_prime_of_degree_ne_zero [IsDomain R] (hf : Prime f
+  lift ((of _).comp f) (root q) <| by simpa [← Polynomial.eval₂_map, ← algebraMap_eq, ← aeval_def]
+
+@[simp] lemma map_of (f : R →+* S) (p : R[X]) (q : S[X]) (h) (r : R) :
+    map f p q h (of p r) = f r := by simp [map]
+
+@[simp] lemma map_root (f : R →+* S) (p : R[X]) (q : S[X]) (h) : map f p q h (root p) = root q := by
+  simp [map]
+
+lemma map_comp_map (f : R →+* S) (g : S →+* T) (p : R[X]) (q : S[X]) (r : T[X]) (hf hg) :
+    (map g q r hg).comp (map f p q hf) =
+      map (g.comp f) p r
+        (hg.trans <| by simpa [Polynomial.map_map] using Polynomial.map_dvd g hf) := by ext <;> simp
+
+/-- `AdjoinRoot.map` as a `RingEquiv`. -/
+def mapRingEquiv (f : R ≃+* S) (p : R[X]) (q : S[X]) (h : Associated (p.map f) q) :
+    AdjoinRoot p ≃+* AdjoinRoot q :=
+  .ofRingHom
+    (map f p q h.symm.dvd)
+    (map f.symm q p <| by
+      -- FIXME: Coercion hell
+      have : (RingHomClass.toRingHom <| (RingEquivClass.toRingEquiv f).symm).comp
+          (RingHomClass.toRingHom f) = .id _ := by ext; exact f.symm_apply_apply _
+      simpa [Polynomial.map_map, this] using map_dvd f.symm.toRingHom h.dvd)
+    (by ext <;> simp) (by ext <;> simp)
+
+@[simp] lemma coe_mapRingEquiv (f : R ≃+* S) (p : R[X]) (q : S[X]) (h) :
+    ⇑(mapRingEquiv f p q h) = map f p q h.symm.dvd := rfl
+
+@[simp] lemma symm_mapRingEquiv (f : R ≃+* S) (p : R[X]) (q : S[X]) (h) :
+    (mapRingEquiv f p q h).symm = mapRingEquiv f.symm q p (by
+      -- FIXME: Coercion hell
+      have : (RingHomClass.toRingHom <| (RingEquivClass.toRingEquiv f).symm).comp
+          (RingHomClass.toRingHom f) = .id _ := by ext; exact f.symm_apply_apply _
+      simpa [Polynomial.map_map, this] using associated_map_map f.symm.toRingHom h.symm) := rfl
+
+variable [CommRing U] [Algebra R S] [Algebra R T] [Algebra R U]
+
+/-- `AdjoinRoot.map` as an `AlgHom`. -/
+def mapAlgHom (f : S →ₐ[R] T) (p : S[X]) (q : T[X]) (h : q ∣ p.map f) :
+    AdjoinRoot p →ₐ[R] AdjoinRoot q where
+  __ := map f p q h
+  commutes' r := by simp [map, AdjoinRoot.algebraMap_eq']
+
+@[simp] lemma coe_mapAlgHom (f : S →ₐ[R] T) (p : S[X]) (q : T[X]) (h) :
+    ⇑(mapAlgHom f p q h) = map f p q h := rfl
+
+lemma mapAlgHom_comp_mapAlghom (f : S →ₐ[R] T) (g : T →ₐ[R] U) (p : S[X]) (q : T[X]) (r : U[X])
+    (hf hg) :
+    (mapAlgHom g q r hg).comp (mapAlgHom f p q hf) =
+      mapAlgHom (g.comp f) p r
+        (hg.trans <| by simpa [Polynomial.map_map] using Polynomial.map_dvd g.toRingHom hf) := by
+  aesop
+
+/-- `AdjoinRoot.map` as an `AlgEquiv`. -/
+def mapAlgEquiv (f : S ≃ₐ[R] T) (p : S[X]) (q : T[X]) (h : Associated (p.map f) q) :
+    AdjoinRoot p ≃ₐ[R] AdjoinRoot q :=
+  .ofAlgHom
+    (mapAlgHom f p q h.symm.dvd)
+    (mapAlgHom f.symm q p <| by
+      -- FIXME: Coercion hell
+      have : (RingHomClass.toRingHom <| (RingEquivClass.toRingEquiv f).symm).comp
+          (RingHomClass.toRingHom f) = .id _ := by ext; exact f.symm_apply_apply _
```

Comment:
What is the plan on fixing this? What is the sequence of rewrites that ideally should closed this goal once we have the refactor done?

Reply:
Can you add a reference to `#21031` in the comment so that we can keep track of these issues and come back to fix it once that is merged?